[upnpcontrol] Add semantic tags

andrewfg opened the pull request:
This PR adds semantic tags.

Note: it is rather tentative since the `Property` tags of some channels are not self evident e.g. 'PlayList' and 'Track' => Perhaps we need to add some more properties?

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
Pinging @mherwege / @jimtng apropos missing properties..

jimtng reviewed (commented):
- I think the following shouldn't be given semantic tags, or at the very least, definitely _not_ `Channel`
  - uri
  - favoriteselect
  - album
  - albumart
  - creator
  - publisher
  - genre
  - tracknumber
  - trackduration
  - trackposition
  - reltrackposition

This should not be set to MediaControl:
- upnprenderer
- browse
- playlistselect
- playlist

jimtng commented:
I don't think we need to apply semantic tags to every single channel. @rkoshak wdyt?

andrewfg commented:
> the following shouldn't be given semantic tags, or at the very least, definitely not Channel

Yeah. I was wondering if we should add a `Track` property ?

jimtng commented:
> Yeah. I was wondering if we should add `Property.Track` ??

If we were to do so, IMO, only _one_ channel should receive this property.
Having multiple channels with the same property seem counter-intuitive and defeats the purpose of semantic tagging.

As I understand it, Semantic tagging is to give meaning to the data, but if you have multiple "items" with the same semantic tag, it is not helpful at all for an external "entity" to "see" and understand your "Thing". If I want to advance the track, which item should I change if there are multiple "Track" channels?

andrewfg commented:
@jimtng you probably misunderstood me. On a player the current `Track` has many sub `Track_Properties` -- Title, Genre, Composer, Artist, Album, Duration, etc..

jimtng commented:
This is how I would've tagged this:

```
<?xml version="1.0" encoding="UTF-8"?>
<thing:thing-descriptions bindingId="upnpcontrol"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">

	<!-- Thing Types -->
	<thing-type id="upnprenderer">
		<label>UPnPRenderer</label>
		<description>UPnP AV Renderer</description>
		<semantic-equipment-tag>MediaPlayer</semantic-equipment-tag>
		<channels>
			<channel id="volume" typeId="system.volume"/>
			<channel id="mute" typeId="system.mute"/>

			<channel id="control" typeId="system.media-control"/>
			<channel id="stop" typeId="stop"/>

			<channel id="repeat" typeId="repeat"/>
			<channel id="shuffle" typeId="shuffle"/>
			<channel id="onlyplayone" typeId="onlyplayone"/>

			<channel id="uri" typeId="uri"/>
			<channel id="favoriteselect" typeId="favoriteselect"/>
			<channel id="favorite" typeId="favorite"/>
			<channel id="favoriteaction" typeId="favoriteaction"/>

			<channel id="playlistselect" typeId="playlistselect"/>

			<channel id="title" typeId="system.media-title"/>
			<channel id="album" typeId="album"/>
			<channel id="albumart" typeId="albumart"/>
			<channel id="creator" typeId="creator"/>
			<channel id="artist" typeId="system.media-artist"/>
			<channel id="publisher" typeId="publisher"/>
			<channel id="genre" typeId="genre"/>
			<channel id="tracknumber" typeId="tracknumber"/>
			<channel id="trackduration" typeId="trackduration"/>
			<channel id="trackposition" typeId="trackposition"/>
			<channel id="reltrackposition" typeId="reltrackposition"/>
		</channels>
		<representation-property>udn</representation-property>
		<config-description>
			<parameter name="udn" type="text" required="true">
				<label>Unique Device Name</label>
				<description>The UDN identifies the UPnP Renderer</description>
			</parameter>
			<parameter name="refresh" type="integer" unit="s">
				<label>Refresh Interval</label>
				<description>Specifies the refresh interval in seconds</description>
				<default>60</default>
			</parameter>
			<parameter name="notificationVolumeAdjustment" type="integer" min="-100" max="100" step="1" unit="%">
				<label>Notification Sound Volume Adjustment</label>
				<description>Specifies the percentage adjustment to the current sound volume when playing notifications</description>
				<default>10</default>
			</parameter>
			<parameter name="maxNotificationDuration" type="integer" unit="s">
				<label>Maximum Notification Duration</label>
				<description>Specifies the maximum duration for notifications, longer notification sounds will be interrupted. O
					represents no maximum duration</description>
				<default>15</default>
			</parameter>
			<parameter name="seekStep" type="integer" min="1">
				<label>Fast Forward/Rewind Step</label>
				<description>Step in seconds for fast forward rewind</description>
				<default>5</default>
			</parameter>
			<parameter name="responseTimeout" type="integer" unit="ms">
				<label>UPnP Response Timeout</label>
				<description>Specifies the timeout in milliseconds when waiting for responses on UPnP actions</description>
				<default>2500</default>
				<advanced>true</advanced>
			</parameter>
		</config-description>
	</thing-type>

	<thing-type id="upnpserver">
		<label>UPnPServer</label>
		<description>UPnP AV Server</description>
		<semantic-equipment-tag>AudioVisual</semantic-equipment-tag>
		<channels>
			<channel id="upnprenderer" typeId="upnprenderer"/>
			<channel id="currenttitle" typeId="system.media-title"/>
			<channel id="browse" typeId="browse"/>
			<channel id="search" typeId="search"/>

			<channel id="playlistselect" typeId="playlistselect"/>
			<channel id="playlist" typeId="playlist"/>
			<channel id="playlistaction" typeId="playlistaction"/>

			<channel id="volume" typeId="system.volume"/>
			<channel id="mute" typeId="system.mute"/>
			<channel id="control" typeId="system.media-control"/>
			<channel id="stop" typeId="stop"/>

		</channels>
		<representation-property>udn</representation-property>
		<config-description>
			<parameter name="udn" type="text" required="true">
				<label>Unique Device Name</label>
				<description>The UDN identifies the UPnP Media Server</description>
			</parameter>
			<parameter name="refresh" type="integer" unit="s">
				<label>Refresh Interval</label>
				<description>Specifies the refresh interval in seconds</description>
				<default>60</default>
			</parameter>
			<parameter name="filter" type="boolean">
				<label>Filter Content</label>
				<description>Only list content which is playable on the selected renderer</description>
				<default>false</default>
			</parameter>
			<parameter name="sortCriteria" type="text">
				<label>Sort Criteria</label>
				<description>Sort criteria for the titles in the selection list and when sending for playing to a renderer. The
					criteria are defined in UPnP sort criteria format. Examples: +dc:title, -dc:creator, +upnp:album. Supported sort
					criteria will depend on the media server</description>
				<default>+dc:title</default>
			</parameter>
			<parameter name="browseDown" type="boolean">
				<label>Auto Browse Down</label>
				<description>When browse or search results in exactly one container entry, iteratively browse down until the
					result
					contains multiple container entries or at least one media entry</description>
				<default>true</default>
			</parameter>
			<parameter name="searchFromRoot" type="boolean">
				<label>Search From Root</label>
				<description>Always search from the root directory</description>
				<default>false</default>
			</parameter>
			<parameter name="responseTimeout" type="integer" unit="ms">
				<label>UPnP Response Timeout</label>
				<description>Specifies the timeout in milliseconds when waiting for responses on UPnP actions</description>
				<default>2500</default>
				<advanced>true</advanced>
			</parameter>
		</config-description>
	</thing-type>

	<!-- Channel Types -->
	<channel-type id="loudness">
		<item-type>Switch</item-type>
		<label>Loudness</label>
		<description>Loudness</description>
		<category>SoundVolume</category>
	</channel-type>
	<channel-type id="stop">
		<item-type>Switch</item-type>
		<label>Stop</label>
		<description>Stop the player</description>
		<autoUpdatePolicy>veto</autoUpdatePolicy>
	</channel-type>
	<channel-type id="repeat">
		<item-type>Switch</item-type>
		<label>Repeat</label>
		<description>Repeat the selection</description>
	</channel-type>
	<channel-type id="shuffle">
		<item-type>Switch</item-type>
		<label>Shuffle</label>
		<description>Random shuffle the selection</description>
	</channel-type>
	<channel-type id="onlyplayone">
		<item-type>Switch</item-type>
		<label>Only Play One</label>
		<description>Stop playback after playing one media entry from queue</description>
	</channel-type>
	<channel-type id="uri">
		<item-type>String</item-type>
		<label>URI</label>
		<description>Now playing URI</description>
	</channel-type>
	<channel-type id="favoriteselect">
		<item-type>String</item-type>
		<label>Select Favorite</label>
		<description>Select favorite to play</description>
		<autoUpdatePolicy>veto</autoUpdatePolicy>
	</channel-type>
	<channel-type id="favorite">
		<item-type>String</item-type>
		<label>Favorite</label>
		<description>Favorite name</description>
	</channel-type>
	<channel-type id="favoriteaction">
		<item-type>String</item-type>
		<label>Favorite Action</label>
		<description>Favorite action</description>
		<command>
			<options>
				<option value="SAVE">Save</option>
				<option value="DELETE">Delete</option>
			</options>
		</command>
		<autoUpdatePolicy>veto</autoUpdatePolicy>
	</channel-type>
	<channel-type id="album">
		<item-type>String</item-type>
		<label>Album</label>
		<description>Now playing album</description>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="albumart">
		<item-type>Image</item-type>
		<label>Album Art</label>
		<description>Now playing album art</description>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="creator">
		<item-type>String</item-type>
		<label>Creator</label>
		<description>Now playing creator</description>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="publisher">
		<item-type>String</item-type>
		<label>Publisher</label>
		<description>Now playing publisher</description>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="genre">
		<item-type>String</item-type>
		<label>Genre</label>
		<description>Now playing genre</description>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="tracknumber">
		<item-type>Number</item-type>
		<label>Track Number</label>
		<description>Now playing track number</description>
		<tags>
			<tag>Status</tag>
			<tag>Track</tag>
		</tags>
		<state readOnly="true"/>
	</channel-type>
	<channel-type id="trackduration">
		<item-type>Number:Time</item-type>
		<label>Track Duration</label>
		<description>Now playing track duration</description>
		<state readOnly="true" pattern="%d %unit%"/>
	</channel-type>
	<channel-type id="trackposition">
		<item-type>Number:Time</item-type>
		<label>Track Position</label>
		<description>Now playing track position</description>
		<state pattern="%d %unit%"/>
	</channel-type>
	<channel-type id="reltrackposition">
		<item-type>Dimmer</item-type>
		<label>Relative Track Position</label>
		<description>Track position as percentage of track duration</description>
		<category>MediaControl</category>
		<tags>
			<tag>Control</tag>
			<tag>Position</tag>
		</tags>
	</channel-type>

	<channel-type id="upnprenderer">
		<item-type>String</item-type>
		<label>Renderer</label>
		<description>Select AV renderer</description>
	</channel-type>
	<channel-type id="browse">
		<item-type>String</item-type>
		<label>Current Media Id</label>
		<description>Current id of media entry or container, option list to browse hierarchy</description>
	</channel-type>
	<channel-type id="search">
		<item-type>String</item-type>
		<label>Search Criteria</label>
		<description>Search criteria for searching the directory. Search criteria are defined in UPnP search criteria format.
			Examples: dc:title contains "song", dc:creator contains "SpringSteen", unp:class = "object.item.audioItem",
			upnp:album contains "Born in"</description>
	</channel-type>
	<channel-type id="playlistselect">
		<item-type>String</item-type>
		<label>Select Playlist</label>
		<description>Playlist for selection</description>
		<autoUpdatePolicy>veto</autoUpdatePolicy>
	</channel-type>
	<channel-type id="playlist">
		<item-type>String</item-type>
		<label>Playlist</label>
		<description>Playlist name</description>
	</channel-type>
	<channel-type id="playlistaction">
		<item-type>String</item-type>
		<label>Playlist Action</label>
		<description>Playlist action</description>
		<command>
			<options>
				<option value="RESTORE">Restore</option>
				<option value="SAVE">Save</option>
				<option value="APPEND">Append</option>
				<option value="DELETE">Delete</option>
			</options>
		</command>
		<autoUpdatePolicy>veto</autoUpdatePolicy>
	</channel-type>
</thing:thing-descriptions>
```

jimtng commented:
We should probably also look at Plex / other media-player stuff like Echo, Chromecast, TVs? Receivers, Sonos?, they'd all have similar stuff

andrewfg commented:
> I'd propose: Point: Track
> Property: Information/Data/Details

I don't think `Track` is a Point.

I would rather suggest `Point.Info` + `Property.TrackInfo` (or `Point.Status` + `Property.TrackInfo` ..)

> We should probably also look at Plex / other media-player stuff 

Agreed. We just need to get it right for the first one, and then it is essentially copy/paste.

jimtng commented:
Point.Status + Property.TrackInfo sounds fine to me.

rkoshak commented:
> I don't think we need to apply semantic tags to every single channel. @rkoshak wdyt?

If it's a Channel that is unlikely to be something exposed to end users of the home OH is controlling it doesn't need a tag. 

But that raises the question of whether the binding author has enough information to make that choice on behalf of the end user and which user should we impose the extra step on, those who do want to expose these to the end user (and therefore they need to select a tag on their own) or those who do not (and therefore they have to remove the auto-suggested tag).

I don't have a good answer. It feels better to always recommend a tag and let the user remove them later. But that's only because of consistency.

Overall, my rule of thumb is only about 60% of Items actually belong in the semantic model. But I'm pretty minimal, so maybe 75-80% for most users. 

> I think the following shouldn't be given semantic tags

I would expect several of those to be something exposed to the end user like artist, album art, etc. If you'd want to see it in MainUI's Overview tabs, you need to tag it with something. 

> Having multiple channels with the same property seem counter-intuitive and defeats the purpose of semantic tagging.

Tags are there to provide a little extra bit of information about what an Item represents. It is not meant to uniquely identify an Item. It's perfectly acceptable to have more than one Channel of a Thing with the same semantic Property tags.

> If I want to advance the track, which item should I change if there are multiple "Track" channels?

The one with the Point tag that indicates it's something you can control. Just like temperature:

- Measurement/Temperature = what's the current temperature
- Setpoint/Temperature = control the desired temperature

So I'd expect something like "trackPosition" to be "Status/Track" or "Position/Track" where "trackDuration" to be "Measurement/Track" or "Status/Track" or something like that. But they are all track Properties. I don't have any problem with having these extra sub-track property tags, but I do want to point out that property tags never stand alone and the tags alone are not there to uniquely identify an Item.

> I don't think Track is a Point.

I strongly agree.  "Track" isn't a verb nor something that can be converted to a verb (in the meaning we are using it here, we are not hunting). It's a noun.

andrewfg commented:
@rkoshak / @jimtng / @jlaur @lsiepel so how shall we proceed here? Whatever we decide here for this binding would set the precedent for all other bindings that play audio or video media. The issues are how to tag channels that inform the state of the "now playing" track or clip. To be specific:

- Tags describing "human" attributes of tracks e.g. Album, Artist, Composer, Title, Genre, Track Number etc. 
- Tags describing "physical" attributes of tracks e.g. Duration, Sample Rate, Channels (mono/stereo/5.1), Medium (mp3, flac) etc.
- Tag(s) describing "state of play" attributes of tracks e.g. Time-position, Time-to-go, Play/Pause etc.

I think we have two options..
1. Tag all such data as `Point.Status` -- eventually in conjunction with (a yet to be defined new) `Property.TrackInfo` (or simply `Property.Info`) tag.
2. Do not tag any such channels.

Personally I am inclined to go for 1. above.

rkoshak commented:
Option 1 makes sense. 2 doesn't bother me either.

mherwege commented:
1 for me. I would stick to Property.Info at first. That can be relevant in non-mediaplayer context as well (e.g. service status messages for car bindings).

andrewfg commented:
@jimtng / @ rkoshak based on feedback from other PRs in the pipeline, I have done the following:

1. applied `Status;Info` to all track metadata channels
2. applied `Control:Progress` to track position controls
3. removed a couple of tags at your suggestion
4. changed some `xyz:MediaControl` to `xyz:Mode`

I am therefore fairly confident that a) this PR is ready to be merged, and b) that it sets a kind of benchmark for other media players still in the pipeline.

lsiepel commented:
When looking at this i find it a bit hard to match the statement from @rkoshak

> Tags are there to provide a little extra bit of information about what an Item represents.

with the tags `Status/Info` . 

To me that is too generic, about all channels in all bindings could be tagged with this.
I would prefer to add more details like `Status/TrackInfo` or maybe even be more specific. like `MusicTrackInfo` or `VideoTrackInfo` ?

andrewfg commented:
TrackInfo was my original suggestion for the new property when we added it to OH core. But others suggested to make it more generic so it could cover other potential use cases.

lsiepel commented:
TrackInfo could also be a child of Info tag. As the info is very specific about music and it is used in many bindings.
I would be interested in core maintainers view on this topic. If you ask me, i would opt for a mroe detailed tag.

andrewfg commented:
..and/or we we could just change the description (far right hand column)

https://github.com/openhab/openhab-core/blob/b94c25a5b847404bb393461aad3fb7794f986aee/bundles/org.openhab.core.semantics/model/SemanticTags.csv#L71

lsiepel commented:
> https://github.com/openhab/openhab-core/blob/b94c25a5b847404bb393461aad3fb7794f986aee/bundles/org.openhab.core.semantics/model/SemanticTags.csv#L71

> https://github.com/openhab/openhab-docs/blob/1bbe162089656e4e37e15458872e77d06aacd668/developers/bindings/thing-xml.md?plain=1#L1027

IMO all should be aligned. 
1. The CSV (very generic text) does not match the XML (specific use described for the tag) 
2. The description from the XML is perfect, but the tag does not match that description. At first the tag should be self-explenatory. Maybe `MediaInfo` is a better match.

i know you might have been told otherwise before, so better that @jimtng, @jlaur or others join the discussion before you adapt anything.

andrewfg commented:
..just for info the word “metadata” is often used to describe what (I think) we mean by the `Info` property..

lsiepel reviewed (approved):
Thanks, LGTM.
